### PR TITLE
Add Duration property to HistoryInfo

### LINKED ISSUE #4181: Add Runtime (Timespan) property to Microsoft.PowerShell.Commands.HistoryInfo
As is, we have StartExecutionTime and EndExecutionTime. 

We should add a property, `RunTime` defined as `EndExecutionTime - StartExecutionTime`, since at least in my experience, I am more concerned with how long a command took to execute than when it ran.

### PR DESCRIPTION
**Pull request by powercode**
Fixes #4181 
Adding a property Duration to HistoryInfo to end the  `{$_.EndExecutionTime - $_.StartExecutionTime}` madness :)

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/powershell/5208)

### CODE REVIEW

**Review by vors — COMMENTED** [review]
sweet! any tests?

**Comment by powercode** [comment]
Not really sure what to test. Testing 
```csharp
public TimeSpan Duration => EndExecutionTime - StartExecutionTime; 
```
 will hardly give a lot. 
But maybe a test that the formatting view is in place?

**Comment by vors** [comment]
> Not really sure what to test

`GetExecutionTimeString` seems like a good candidate for unit tests

**Comment by iSazonov** [comment]
@powercode Please resolve merge conflicts.

**Comment by powercode** [comment]
@vors You were so right. Added tests and fixed bugs

**Review by vors — APPROVED** [review]
Looks solid, but not a big fan of reflection in the tests. This is very fragile.
I guess it should not be a blocker, but maybe we can just do some unit testing from C#? Don't we have such tests?

File: src/System.Management.Automation/engine/hostifaces/History.cs [hunk 1]
```diff
@@ -36,11 +34,11 @@ internal HistoryInfo(long pipelineId, string cmdline, PipelineState status, Date
         {
             Dbg.Assert(cmdline != null, "caller should validate the parameter");
             _pipelineId = pipelineId;
-            _cmdline = cmdline;
-            _status = status;
-            _startTime = startTime;
-            _endTime = endTime;
-            _cleared = false;
+            CommandLine = cmdline;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs]: @SteveL-MSFT @PaulHigin Does this not break remoting?

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 [hunk 2]
```diff
@@ -108,4 +108,62 @@
...
         $errorResult | Should be CommandNotFoundException
     }
+
+    It "Has time view in table format" {
+        $null = Get-ChildItem
+        Get-History | Format-Table -View time -ErrorVariable ftError | Out-Null
+        $ftError.Count | Should Be 0
+    }
+
+    It "HistoryInfo calculates Duration" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: These tests are very similar. Please use `-TestCases`.

**Review by iSazonov — COMMENTED** [review]
Leave a comment

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 (hunk 2 above)
Comment (powercode) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: good point

File: src/System.Management.Automation/engine/hostifaces/History.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/hostifaces/History.cs]: There's a specific [list of types](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/remoting/common/WireDataFormat/RemoteHostEncoder.cs#L41) and HistoryInfo isn't in that list

File: src/System.Management.Automation/engine/hostifaces/History.cs [hunk 3]
```diff
@@ -195,54 +147,31 @@ internal void SetEndTime(DateTime endTime)
...
-        /// Flag indicating an entry is present/cleared
-        /// </summary>
-
-        private bool _cleared = false;
-
-        #endregion private
-
-        #region ICloneable Members
+        /// <returns></returns>
+        public string GetDurationString()
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/History.cs]: Minor comment - should we follow `ToString()` and use `ToDurationString()` for public?

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 [hunk 4]
```diff
@@ -108,4 +108,44 @@
...
+        $ctor = [Microsoft.PowerShell.Commands.HistoryInfo].GetConstructors([Reflection.BindingFlags]::NonPublic -bor [Reflection.BindingFlags]::Instance).Where{$_.GetParameters().Length -eq 5}
+        $start = [datetime]::new(2001, 01, 01, 10, 01, 01)
+        $duration = [timespan] "1:2:21"
+        $end = $start + $duration
+        $hi = $ctor.Invoke((1, "command", [Management.Automation.Runspaces.PipelineState]::Completed, $start, $end))
+        $hi.Duration | Should Be $duration
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: Should we remove the test?

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 [hunk 5]
```diff
@@ -108,4 +108,44 @@
...
+    }
+
+    It "Displays HistoryInfo with Duration <expectedResult> for timespan <duration>" -TestCases @(
+        @{duration = "0:0:21"; expectedResult = "00:21.000"}
+        @{duration = "1:14:21"; expectedResult = "01:14:21"}
+        @{duration = "1.1:14:21"; expectedResult = "1.01:14:21"}
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: `TimeSeparator` can be culture sensitive. Could you please use TimeSpan constructor for `duration`?

Also in C# `$@"{duration:d\.hh\:mm\:ss}";` should we be culture sensitive too?

**Review by iSazonov — COMMENTED** [review]
Leave a comment

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 (hunk 5 above)
Comment (powercode) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: @iSazonov Excellent catch!
I will fix!

File: src/System.Management.Automation/engine/hostifaces/History.cs (hunk 3 above)
Comment (powercode) [reply src/System.Management.Automation/engine/hostifaces/History.cs]: Yes, or `Tostring(string)`. Thought about that a bit also, but never reach any real conclusion.

**Comment by SteveL-MSFT** [comment]
We're not holding the RC for this PR, if it makes it in soon, great!  If not, it'll catch the next release.

**Comment by mi-hol** [comment]
@SteveL-MSFT will it merge this time?

**Comment by SteveL-MSFT** [comment]
@mi-hol if the maintainer @iSazonov believes there has been sufficient review, he can merge this, but we're limiting code changes we're taking for GA to avoid risk of regression.  This will show up in 6.1.0 preview release which should happen early next year.

**Comment by iSazonov** [comment]
We have only one open comment about `TimeSeparator`. 
I don't see any reason why we won't get it in 6.1.0.

**Comment by powercode** [comment]
I have looked a bit on how PowerShell treats timespans in other cases, and it seems to use the default formatting, which is not culture sensitive.
I think we should stick with consistency here, and keep the implementation as it is.

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 (hunk 5 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: Closed.

**Comment by iSazonov** [comment]
@powercode please resolve conflicts.

**Comment by iSazonov** [comment]
@SteveL-MSFT @daxian-dbw Have you any thoughts about the PR?

**Review by SteveL-MSFT — DISMISSED** [review]
Regarding `GetDurationString()` vs `ToDurationString()`, looking in PS code, it seems `Get*String()` is used (although internal methods).  I'm fine keeping `GetDurationString()` unless someone can point to some guidance otherwise.

**Comment by iSazonov** [comment]
@TravisEz13 @daxian-dbw Have you any thoughts about `GetDurationString()` vs T`oDurationString()`?

**Comment by TravisEz13** [comment]
`ToDurationString()`seems more appropriate, but I don't find either wrong.

**Comment by iSazonov** [comment]
@SteveL-MSFT Please make conclusion and the public API should be approved by PowerShell Committee.

**Comment by SteveL-MSFT** [comment]
@iSazonov Committee meetings are every Wed and we review PRs first so we'll have an answer tomorrow

**Comment by TravisEz13** [comment]
@SteveL-MSFT  what was the committee answer?

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and recommends that `GetDurationString()` is not needed as `[TimeSpan]` has a `ToString()` implementation already which is consistent regardless of the duration which is preferred.  It would have been great to have an example of what the output would look like in the PR description where format/output is involved.

**Comment by powercode** [comment]
It was just an attempt to scale the output to a range that is most commonly used.

But I was never sure it was the right thing to do anyways. And it can however easily be added as a formatter by those who so choose.

**Comment by rkeithhill** [comment]
I'd love to see the property added and yeah, let folks format the TimeSpan differently if they want.  I could even see some wanting "humanizer" versions of the duration which they could configure with format data - as you point out.

**Comment by bergmeister** [comment]
@powercode Please upstream and resolve merge conflict.

**Comment by powercode** [comment]
These build errors are not related to my changes.

Can anyone restart the build here?

**Comment by iSazonov** [comment]
Reopen the PR to restart CIs.

**Comment by iSazonov** [comment]
@powercode Is the PR ready for review?

**Comment by powercode** [comment]
Yes

**Comment by iSazonov** [comment]
@daxian-dbw @SteveL-MSFT @BrucePay Please review the PR.

**Review by TravisEz13 — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/powershell/powershell/5208)* status: :shipit: complete! all files reviewed

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 [hunk 6]
```diff
@@ -110,4 +110,13 @@ Describe "History cmdlet test cases" -Tags "CI" {
 
         $errorResult | Should -BeExactly 'CommandNotFoundException'
     }
+
+    It "HistoryInfo calculates Duration" {
+        $ctor = [Microsoft.PowerShell.Commands.HistoryInfo].GetConstructors([Reflection.BindingFlags]::NonPublic -bor [Reflection.BindingFlags]::Instance).Where{$_.GetParameters().Length -eq 5}
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: Seems like this test should just get the first item from item from `Get-History` and validate that Duration = End - Start?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 (hunk 6 above)
Comment (powercode) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: @SteveL-MSFT  Am I guaranteed to have something in the history at that moment?

Comment (powercode) [reply test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: Fixed

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

File: test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1 [hunk 7]
```diff
@@ -110,4 +110,19 @@ Describe "History cmdlet test cases" -Tags "CI" {
...
+        $end = $start + $duration
+        $history = [PSCustomObject] @{
+                CommandLine="command"
+                ExecutionStatus=[Management.Automation.Runspaces.PipelineState]::Completed
+                StartExecutionTime=$start
+                EndExecutionTime=$end
+        }
+        $history | Add-History
+        $h = Get-History -count 1
+        $h.Duration | Should Be $duration
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/History.Tests.ps1]: @powercode Sorry I lost the PR. Please fix `Be` -> `-Be` and we ready to merge.

**Comment by iSazonov** [comment]
@powercode Thank you for the contribution! Sorry for delay.

**Pull request merged**